### PR TITLE
Fix Video reader race condition

### PR DESCRIPTION
**Pull request by JanuszL**
- there is a race condition between the video reader
  destructor and the reader Run method. After the reader
  run method, all current batch samples are zeroed
  to recycle the underlying storage. However, the video
  reader destructor calls the` wait` method on all samples
  in the prefetch buffer to make sure that all underlying
  kernels that operate on them ended. There is a short period
  between zeroing samples and creating a fresh batch where
  the destructor may attempt to access nullptr. This PR fixes
  the issue but not only clearing all samples but reducing
  vector length that holds them to 0 so the destructor won't
  access them as it iterates only over available samples in the vector

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

## Description
- [x] **Bug fix** (*non-breaking change which fixes an issue*)
- [ ] **New feature** (*non-breaking change which adds functionality*)
- [ ] **Breaking change** (*fix or feature that would cause existing functionality to not work as expected*)
- [ ] **Refactoring** (*Redesign of existing code that doesn't affect functionality*)
- [ ] **Other** (*e.g. Documentation, Tests, Configuration*)

### What happened in this PR
- there is a race condition between the video reader
  destructor and the reader Run method. After the reader
  run method, all current batch samples are zeroed
  to recycle the underlying storage. However, the video
  reader destructor calls the` wait` method on all samples
  in the prefetch buffer to make sure that all underlying
  kernels that operate on them ended. There is a short period
  between zeroing samples and creating a fresh batch where
  the destructor may attempt to access nullptr. This PR fixes
  the issue but not only clearing all samples but reducing
  vector length that holds them to 0 so the destructor won't
  access them as it iterates only over available samples in the vector

#### Additional information
- Affected modules and functionalities:
  - reader_op.

- Key points relevant for the review:
  - NA 

### Tests
- [x] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-2292

### CODE REVIEW

**Comment by JanuszL** [comment]
!build

**Review by mzient — APPROVED** [review]

File: dali/operators/reader/reader_op.h [hunk 1]
```diff
@@ -179,9 +179,8 @@ class DataReader : public Operator<Backend> {
     // Consume batch
     Operator<Backend>::Run(ws);
     CUDA_CALL(cudaStreamSynchronize(ws.stream()));
-    for (int sample_idx = 0; sample_idx < max_batch_size_; sample_idx++) {
-      auto sample = MoveSample(sample_idx);
-    }
+    // clear samples from the current batch so they can be recycled
+    ClearCurentBatch();
```
Comment (mzient) [inline dali/operators/reader/reader_op.h:183]: So, after `Run` this queue entry is empty, right? It's now the responsibility of `Prefetch` to resize it to the right size.
While you're at it, could you swap `reserve` and `clear` in `Prefetch`? Now `reserve` goes before `clear` which is rather counterproductive.

**Review by mzient — APPROVED** [review]

File: dali/operators/reader/reader_op.h (hunk 1 above)
Comment (JanuszL) [reply dali/operators/reader/reader_op.h:183]: > So, after Run this queue entry is empty, right?

Yes

> It's now the responsibility of Prefetch to resize it to the right size.

It always was but was used only during the first iterations. Now it happens always.

> While you're at it, could you swap reserve and clear in Prefetch? Now reserve goes before clear which is rather counterproductive.

Done.

**Comment by JanuszL** [comment]
!build

**Review by stiepan — APPROVED** [review]

**Pull request merged**